### PR TITLE
Add definition of initial table entries, without const qualifier

### PR DESCRIPTION
**Pull request by jfingerh**
Summary of comments with approve/disapprove/change-requests as of 2022-Dec-27:

+ Sander Tolsma approved on 2022-Nov-16 after several of his review comments were addressed.
+ Mario Baldi approved in 2022-Nov
+ Steve Licking  approved in 2022-Nov
+ Mihai Budiu wrote several comments and created partial implementation in p4c.  Not clear whether he approves yet.
+ Andrew Pinski wrote some comments.  Overall approves of the idea, but would prefer a different syntax for specifying priorities of entries.

@thantry Please do provide your review comments, or approval of this PR as written.

### CODE REVIEW

**Comment by jfingerh** [comment]
Intended to be a change to the spec that resolves issue https://github.com/p4lang/p4-spec/issues/1159, but note that it is not complete yet.  You are free to add review comments for the changes that have been made so far, but the section defining the options for initial entry priorities has not been written yet.

File: p4-16/spec/P4-16-spec.mdk
Comment (stolsma) [inline p4-16/spec/P4-16-spec.mdk:6943]: How is this convention called?? (the P4Runtime strategy is called 'largest-priority_wins', so I'm curious what is this called)

File: p4-16/spec/P4-16-spec.mdk
Comment (stolsma) [inline p4-16/spec/P4-16-spec.mdk:6941]: > non-negative integers

Also same question: Defined as?? Starting at and including 0 (zero) ?

File: p4-16/spec/P4-16-spec.mdk
Comment (jfingerh) [reply p4-16/spec/P4-16-spec.mdk:6936]: In mathematics, positive usually means strictly greater than 0, so starting at 1.  Non-negative means not less than 0, so starting at 0, but I'm happy to clarify both of these in the text if it will help.

File: p4-16/spec/P4-16-spec.mdk
Comment (jfingerh) [reply p4-16/spec/P4-16-spec.mdk:6943]: Well largest_priority_wins is just the name I'm using in this document, to give a name to it.  I would call TDI smallest_priority_wins.

File: p4-16/spec/P4-16-spec.mdk
Comment (mbudiu-vmw) [inline p4-16/spec/P4-16-spec.mdk:6921]: How do you specify the initial priorities?

Comment (jfingerh) [reply p4-16/spec/P4-16-spec.mdk:6921]: To be described in the part of the PR that I have not yet written.  Probably another week or two.

Comment (jfingerh) [reply p4-16/spec/P4-16-spec.mdk:6937]: @mbudiu-vmw I added another commit to this PR on 2022-Nov-15 with a description of how priorities are determined.

**Comment by jfingerh** [comment]
@antoninbas @smolkaj @jnfoster @thantry @mariobaldi @mbudiu-vmw This is ready for review.  Corrections and comments (or approval) before the next P4 LDWG meeting on 2022-Dec-05 would be much appreciated, if you have time.  Otherwise it probably won't get into the spec for another month or two after that (if it gets into the spec).

Note that the per-entry `const` qualifier and per-entry initial priority values might require extensions to the data model of P4Runtime API, e..g perhaps adding an optional `const` field for each entry, and the auto-generated `.entries.txt` would need a way to specify that a table had initial entries without a global `const` before `entries`.

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [inline p4-16/spec/P4-16-spec.mdk:6852]: The only reason to change the name of `a_with_control_params` in this example is to make it shorter, so I can reuse most of this example in a new one, and have more page width to work with for adding priority values and comments.

**Comment by jafingerhut** [comment]
@vgurevich @sayanb @slicking This is ready for review. Corrections and comments (or approval) before the next P4 LDWG meeting on 2022-Dec-05 would be much appreciated, if you have time. Otherwise it probably won't get into the spec for another month or two after that (if it gets into the spec).

See my previous comment on this PR, regarding new fields likely need for P4Runtime `.entries.txt` files.  This feature addition seems likely to require similar enhancements in TDI/bfrt.

File: p4-16/spec/P4-16-spec.mdk
Comment (stolsma) [reply p4-16/spec/P4-16-spec.mdk:6952]: @jfingerh thanks for the clarifying text! Really nice!! 👍 A long time  ago I went to a meteorology study and there the rule was everything upwards -0.0 is called a positive temperature (I know, the '-' is crazy) so from that moment on I always ask for precise definition of positive etc...

**Review by stolsma — APPROVED** [review]

**Comment by jafingerhut** [comment]
@thantry @antoninbas @smolkaj @mariobaldi @vgurevich @sayanb @slicking  This is ready for review. Corrections and comments (or approval) before the next P4 LDWG meeting on 2022-Dec-05 (one week from today) would be much appreciated, if you have time. Otherwise it probably won't get into the spec for another month or two after that (if it gets into the spec).

Note that the per-entry const qualifier and per-entry initial priority values might require extensions to the data model of P4Runtime API, e..g perhaps adding an optional const field for each entry, and the auto-generated .entries.txt would need a way to specify that a table had initial entries without a global const before entries.

This feature seems likely to require similar enhancements in TDI/bfrt.

**Comment by mariobaldi** [comment]
This looks good to me. I am in favor of approving this.

File: p4-16/spec/P4-16-spec.mdk
Comment (mbudiu-vmw) [reply p4-16/spec/P4-16-spec.mdk:6822]: This will forbid using expressions here, e.g., constant values.

**Review by mbudiu-vmw — CHANGES_REQUESTED** [review]
See a working implementation here: https://github.com/p4lang/p4c/pull/3748

**Comment by mbudiu-vmw** [comment]
@jafingerhut please note that the grammar requires more extensive changes to allow "priority" to be used as a regular identifier.
(We should review the spec grammar again, it may have fallen behind the implementation.)

**Review by slicking — APPROVED** [review]
This looks good to me, thanks @jfingerh for working on this.  I remember requests to add additional entries at runtime to a table with initial entries has come up before.  So this will be a nice addition.  Looks very complete and I'm also glad to see the ability to mix const and non-const entries here.

File: p4-16/spec/P4-16-spec.mdk
Comment (jfingerh) [reply p4-16/spec/P4-16-spec.mdk:7115]: We do have a `@priority` annotation already implemented in p4c, yes.  The reason for the suggested syntax NOT using that here is the same general "prefer NOT to use annotations for things that can change the semantics of the program if they were removed" guidance that we often hear in the LDWG meetings when annotations are proposed.

File: p4-16/spec/P4-16-spec.mdk
Comment (mbudiu-vmw) [inline p4-16/spec/P4-16-spec.mdk:6844]: I don't know if it's worth specifying that having an entry in a table that is unreachable is possible, so the fact something can't be removed doesn't mean it will ever apply.

File: p4-16/spec/P4-16-spec.mdk
Comment (mbudiu-vmw) [inline p4-16/spec/P4-16-spec.mdk:6931]: I guess we have to merge https://github.com/p4lang/p4-spec/pull/1145 first

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [reply p4-16/spec/P4-16-spec.mdk:7115]: There is a `@priority` annotation implemented in p4c since a 2017 commit to p4c by Calin Cascaval, but it is described nowhere in the language spec, and I believe never has.

Do you want me to deprecate something not mentioned in the spec, in the spec?

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [reply p4-16/spec/P4-16-spec.mdk:6844]: By "unreachable" do you mean that it is impossible that a table.apply() call could ever find it as the "winning entry", because other higher priority entries mask it?

If so, this concept is never mentioned in the spec, and as far as I can see, need not be mentioned in the spec.

If an implementation wants to optimize away such entries, that seems to me a fine optimization that an implementation could make, but the spec doesn't need to mandate or forbid it.  The spec defines the observable behavior.

File: p4-16/spec/P4-16-spec.mdk
Comment (mbudiu-vmw) [reply p4-16/spec/P4-16-spec.mdk:7115]: Maybe you can find out if the barefoot backend uses them. We care about implementations even if they are not conformant with the spec.

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [reply p4-16/spec/P4-16-spec.mdk:6931]: We could merge #1145 first, but I do not see any way in which this PR depends upon changes made in #1145.  Do you?

Comment (mbudiu-vmw) [reply p4-16/spec/P4-16-spec.mdk:6931]: What each of these kinds does is described there

Comment (jafingerhut) [reply p4-16/spec/P4-16-spec.mdk:6931]: I guess I would prefer if merging this PR was not blocked waiting for #1145, if #1145 is judged as not ready to merge for some reason, but this one is.  I think the text of this PR makes perfect sense even in the absence of #1145 being merged in, even if it might be more fully understood by someone first encountering what exact, lpm, and ternary mean after reading #1145's changes.  But the spec has lived for 6 years without #1145, so it doesn't seem like that should be blocking other enhancements now.

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [reply p4-16/spec/P4-16-spec.mdk:6807]: I will compare your grammar changes in this p4c PR against what I wrote in this PR, and get back with updates to this PR, and/or questions about the differences in future comments.

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [reply p4-16/spec/P4-16-spec.mdk:6822]: Good point on the forbidding of expressions.  I will think about what grammar would enable expressions, but whatever expressions we allow, they should evaluate at compile time to integer constants.

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [reply p4-16/spec/P4-16-spec.mdk:7115]: Checking with @slicking @sayanb @vgurevich to see if they know.

File: p4-16/spec/P4-16-spec.mdk
Comment (mbudiu-vmw) [reply p4-16/spec/P4-16-spec.mdk:6822]: The implementation supports expressions, but they must be parenthesized to be unambiguous.
Similar to the `bit<(expression)>` case.

File: p4-16/spec/P4-16-spec.mdk
Comment (apinski-cavium) [inline p4-16/spec/P4-16-spec.mdk:6826]: Hmm, an entry would look like
`const priority = 1     2 ... 3   :   a(t)    @opn;`
I don't know if like 1 and 2 being next to each other there without some kind of separator. 
The P4 language has separators for most things, having a separator removes the ambiguous issue too and why the need () for the expression case too.
I am ok with not having a separator but many folks might have to do a double take of looking at the code to understand it.

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [reply p4-16/spec/P4-16-spec.mdk:7115]: I checked with @slicking and the only uses he found of `@priority` in our code were test programs copied from public p4c.

Comment (mbudiu-vmw) [reply p4-16/spec/P4-16-spec.mdk:7115]: In fact, this is what the compiler outputs if you use `@priority`:
```
table-entries-priority-bmv2.p4(79): [--Wwarn=deprecated] warning: The @priority annotation on Entry is not part of the P4 specification, nor of the P4Runtime specification, and will be ignored
            0x1181 &&& 0xF00F : a_with_control_params(3) @priority(1);
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

So I guess they are already deprecated.

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [reply p4-16/spec/P4-16-spec.mdk]: With commit 9 I fixed priority->PRIORITY and also added the grammar rule to allow parenthesized expressions as priority values.  Thanks.

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [reply p4-16/spec/P4-16-spec.mdk]: With commit 10 on this PR I have updated the grammar in the spec to match what p4c has, at least for the non-terminal symbol tableProperty

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [reply p4-16/spec/P4-16-spec.mdk:6825]: I agree that does not look great.

Note that the syntax also allows one to write it more clearly as:
```
const priority = 1 (2 .. 3) : a(t) @opn;
```
We could recommend that people use the parentheses there for clarity, perhaps?

File: p4-16/spec/P4-16-spec.mdk
Comment (mbudiu-vmw) [inline p4-16/spec/P4-16-spec.mdk:7130]: these priorities do not change when the table is modified.

Comment (jfingerh) [reply p4-16/spec/P4-16-spec.mdk:7133]: I added a short paragraph to the text explaining this.  Rather than saying that the priorities do not change when the table is modified, I left it up to the behavior defined by the control plane API in use.  This leaves open the possibility for a control plane API to define mechanisms to modify the priority values if entries, if this is deemed useful, without contradicting the language spec.

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [reply p4-16/spec/P4-16-spec.mdk:6825]: I agree we could.  I do not have a strong preference either way.  By what decision process should this syntax question be decided?

**Comment by jafingerhut** [comment]
TODO for Andy when all other open questions about this PR have been resolved: Add changes to grammar.mdk file, and consider whether `priority` should be added to list of keywords or not.

**Comment by apinski-cavium** [comment]
Overall I am ok with this change. I am not happy with the syntax around priorities but I won't block the change if other people are ok with the syntax.

**Comment by jfingerh** [comment]
@apinksi-cavium Any alternate syntax(es) for specifying priorities that you might suggest?

**Comment by apinski-cavium** [comment]
> @apinksi-cavium Any alternate syntax(es) for specifying priorities that you might suggest?

Just have a colon (':') after the priority if it is present. But as I said I don't want to bike shed the syntax here (https://en.wikipedia.org/wiki/Law_of_triviality ).

**Comment by jfingerh** [comment]
Understood the desire not to bikeshed, but if we don't have comments here about it, the only other place is the LDWG meetings, and they are short enough that I don't want to wait until those meetings to resolve issues, if possible.

**Comment by jnfoster** [comment]
If we are brainstorming syntax, how about using curly braces for an entry, with the priority before it?

```p4
...
  entries = [
    10 : { 1, a1() },   // priority 10
    { 2,  a2(99) },   // no priority given; compiler allocates 
    _ : { 3, a3() },   // priority explicitly elided, compiler allocates
    100 : { 4, a4() }, // priority 100
    200 : { 5 & 1, a5() } // priority 200
] 
```
I like the idea that an entry (match + action) is a unit of data that gets special syntactic treatment. I prefer having the priority outside of the entry. But one could also consider moving it inside...

As a further extension, one could also allow using optional labels if being extra explicit:
```p4
  100 : { key = 4, action = a4() }, 
```

And we could support both 
```
entries = { ... } // no order implied
```
and 
```
entries = [ ... ] // order implied
```
which matters when the compiler is allocating priorities.

This syntax is obviously not backwards compatible, but we could deprecate the earlier syntax for `const entries` in several versions. I don't think it's in widespread use, so the disruption should be minimal...

**Comment by thantry** [comment]
I like Nate’s way of specification. It starts to look more like a hashtable with composite types.Otherwise, looks quite good to me.ThanksHariSent from my iPhoneOn Dec 28, 2022, at 9:40 PM, Nate Foster ***@***.***> wrote:﻿
If we are brainstorming syntax, how about using curly braces for an entry, with the priority before it?
...
  entries = [
    10 : { 1, a1() },   // priority 10
    { 2,  a2(99) },   // no priority given; compiler allocates 
    _ : { 3, a3() },   // priority explicitly elided, compiler allocates
    100 : { 4, a4() }, // priority 100
    200 : { 5 & 1, a5() } // priority 200
] 
I like the idea that an entry (match + action) is a unit of data that gets special syntactic treatment. I prefer having the priority outside of the entry. But one could also consider moving it inside...
As a further extension, one could also allow using optional labels if being extra explicit:
  100 : { key = 4, action = a4() }, 
And we could support both
entries = { ... } // no order implied

and
entries = [ ... ] // order implied

which matters when the compiler is allocating priorities.
This syntax is obviously not backwards compatible, but we could deprecate the earlier syntax for const entries in several versions. I don't think it's in widespread use, so the disruption should be minimal...

**Comment by jafingerhut** [comment]
Hari wrote: "I like Nate’s way of specification. It starts to look more like a hashtable with composite types."

I have no objections to the syntax.  I would like to caution one about explaining it as a hash table, though.  Both the P4Runtime API and TDI control plane APIs explicitly support and allow there to be multiple entries in a table with the same priority as each other.

Background details on what it means if multiple entries have the same priority, and why it might be useful: Which one is the "winner" if more than one entry with the same numeric priority is not specified by the P4Runtime API.  Some control plane software only uses this capability in situations where all entries with the same priority are mutually exclusive, i.e. at most one of them can match any given search key value.  It can provide efficiency advantage when adding entries to such tables, in that fewer entries need to be moved when new entries are added.

**Review by mbudiu-vmw — APPROVED** [review]

**Comment by mbudiu-vmw** [comment]
@jafingerhut will this be ready for discussion at the next LDWG meeting?

File: p4-16/spec/P4-16-spec.mdk
Comment (jafingerhut) [reply p4-16/spec/P4-16-spec.mdk:6930]: #1145 is merged now, so resolving this comment.

**Comment by jafingerhut** [comment]
Sorry, still no updates to this PR from me, but I would like to try to nail down the desired syntax, if possible.

This comment probably uses too many words to explain why I think we don't need to introduce two separate delimiters like square brackets vs. curly brackets for a list of table entries, and why curly brackets should be sufficient.

Gory details:

Today's `const entries = { .... }` syntax DOES imply an order to the entries in the `...` list.  See this paragraph in the current language spec:

```
If the runtime API requires a priority for the entries of a table—e.g. when using the P4 Runtime API, tables with at least one ternary search key field—then the entries are matched in program order, stopping at the first matching entry. Architectures should define the significance of entry order (if any) for other kinds of tables.
```

I mention this because of the proposal in comments above for a syntax using square brackets around a list of entries, e.g. `entries = [ ... ]`, with order required, vs. `{ ... }` with order not required.  It seems to me like this would be a backwards-incompatible change, and while we can certainly consider such a backwards-incompatible change, we should do it fully aware that we are doing so, and that it is worth doing so.

It is not clear to me that having a way to specify "order not required" is useful here.  Sure, there are tables where all key fields are exact match_kind only, and there the order is not important as long as all key values are distinct (and I believe we should require them to be distinct, or give an error message at compile time, in for such exact match tables).  But having only a syntax that means "source code of entries is significant" is easy to implement in that case: the compiler can ignore that order if it wants to.

**Comment by jafingerhut** [comment]
Alternate syntax for initial entries has been proposed in this alternate PR.  At most one of this PR and that one should be merged: https://github.com/p4lang/p4-spec/pull/1231

**Comment by jafingerhut** [comment]
@jnfoster @thantry I have tried to implement a grammar similar to what you propose in your comments above, but everything I have tried that looks similar to your example syntaxes leads to a grammar that causes shift/reduce conflicts in bison.  My work in progress version of it can be found on an alternate PR here: #1231 

Unless someone can find a bison-friendly grammar for an alternate syntax they like, that alternative syntax will not make any progress.

**Comment by jafingerhut** [comment]
Commits 13 & 14 made on 2023-Apr-10 should make no changes to this PR's content, other than to merge it with the latest main branch version of the spec, which had some conflicts because of the use of `[INCLUDE=grammar.mdk:nonTerminalName]` Madoko constructs in the latest main version, which was merged into main after this PR was first written.

**Comment by jafingerhut** [comment]
Current summary of this issue and possible proposed syntaxes for specifying the initial entries:

See slides 35 and 36 near the end of this slide deck: https://docs.google.com/presentation/d/1GW-9FDtyLuwqV_mdGeDh_qqDiey0VkQ1/edit#slide=id.g22d0e434be1_0_5

It shows examples of the syntaxes of the two PRs that exist for this issue, with links to those two PRs.

I am pinging people who have given comments on preferred syntaxes here.  I am personally happy with any of the ones mentioned in the slides linked above.  If someone wants to come up with another syntax that is not one of those, then it almost certainly will not happen in the P4 spec release in April 2023, but some later month after we discuss it again in a later LDWG meeting.

@mbudiu-vmw @jnfoster @thantry @apinski-cavium @vgurevich

**Comment by thantry** [comment]
Approve! Thanks for the great work in pushing this through!

**Review by jnfoster — APPROVED** [review]

**Review by jnfoster — APPROVED** [review]

**Pull request merged**
